File: MaxCounters.cxx
#include <vector>

using namespace std;

vector<int> solution(int N, vector<int> &A) {

    // stores the returned answer (this automatically
    // initializes all N counters to zero) 
    vector<int> answer(N);


    // keep track of the last counter maxout value
    // so that we don't have to loop over every counter
    // each time the counters need to be set to
    // the maximal value among them
    int M = 0;

    // current counter maximum value at any given iteration through A
    int m = 0;

    // look at each element of A and perform corresponding
    // operations on counters
    for (vector<int>::iterator it = A.begin(); it != A.end(); it++) 
    {
        // counter number that's being referenced by the array's entry
        int& cntnum = (*it);

        // increment the corresponding counter if the counter number value
        // is in 1 to N range
        if(1 <= cntnum && cntnum < N + 1) 
        {
            
            // counter that's being incremented
            int& cntval = answer[cntnum - 1];

            // setting the counter to the last maxout value 
            // if it's below that value
            if(cntval < M)
                cntval = M; 

            // increment the current counter's value
            cntval ++;

            // and keeping the current maximum value
            if(cntval > m)
                m = cntval;
        }
        // if the counter number is at N + 1 then
        // set the maxout value to the current maximum
        else if (cntnum == N + 1)
            M = m;
        // return an emtpy answer if the counter number is wrong
        else
            return vector<int>(0);
    }

    // be sure that all counters have been set to the last maxout value
    if(M > 0)
        for (vector<int>::iterator it = answer.begin(); it != answer.end(); it++)
        {
            int& cntval = (*it);
            if(cntval < M)
                cntval = M;
        }

    // return the answer
    return answer;
}

File: MaxCounters.py

# present problem doesn't allow numpy
#import numpy

def solution(N, A):
    counters=[int(0)]*N
    m = int(0) # maximum value at each iteration through A
    M = int(0) # last maximum value all counters were set to
    # iterate through the elements of A
    for cntnum in A:
        # instruction to increment a certain counter
        if 1 <= cntnum and cntnum < N + 1:
            # make sure that the last maxout value was
            # taken into account
            if counters[cntnum - 1] < M:
                counters[cntnum - 1] = M
            # increment the counter with index corresponding 
            # to the present value of A
            counters[cntnum - 1] += 1
            # update the current maximum value across counters
            if m < counters[cntnum - 1]:
                m = counters[cntnum - 1]
        # instruction to set all counters to the maximum value
        # across all counters
        elif cntnum == N + 1:
            M = m # update the last maxout value
        # wrong A value, return None when this happens
        else:
            return None
    # set all counters that are less than the last maxout
    # value to the maxout value and return the answer
    return [ M if c < M else c for c in counters]
